### PR TITLE
Bruteforce protection implementation

### PR DESCRIPTION
**Pull request by mihaidma**
Implements: https://github.com/senecajs/seneca-user/issues/62

### CODE REVIEW

File: test/lock-unlock.test.js [hunk 1]
```diff
@@ -1,148 +1,184 @@
...
-si.use('../user', {lockTry: 3})
 
+var user_canon = 'sys/user'
 var user1Data = {
   nick: 'nick1',
   email: 'nick1@example.com',
   password: 'test1test',
-  active: true
+  active: true,
+  id: ''
```
Comment (mirceaalexandru) [inline test/lock-unlock.test.js]: I think id could be ignored - not specified here

File: test/lock-unlock.test.js [hunk 2]
```diff
@@ -1,148 +1,184 @@
...
 
-  test('user/register test', function (done) {
-    si.act(_.extend({role: 'user', cmd: 'register'}, user2Data), function (err, data) {
+        next()
+      })
+    }, function (err, results) {
       expect(err).to.not.exist()
-      expect(data.user.nick).to.equal(user2Data.nick)
-      done(err)
+      done()
```
Comment (mirceaalexandru) [inline test/lock-unlock.test.js]: Here it would be great to test if a correct login work as expected or not.

File: test/lock-unlock.test.js [hunk 3]
```diff
@@ -1,148 +1,184 @@
...
-    })
-  }
-
-  test('user/login user test', function (done) {
-    si.act({role: 'user', cmd: 'login', nick: user1Data.nick, password: user1Data.password}, function (err, data) {
+    }, function (err, results) {
       expect(err).to.not.exist()
-      expect(data.ok).to.be.false()
-      done(err)
+      done()
```
Comment (mirceaalexandru) [inline test/lock-unlock.test.js]: Same - here test if a correct login works or not - in this case should not work

File: user.js [hunk 4]
```diff
@@ -633,7 +633,7 @@ module.exports = function user (options) {
       return done(null, {ok: false, why: why, user: user})
     }
 
-    if (user.lockTry >= options.lockTry && !_.isNull(options.lockTry)) {
+    if (options.failedLoginCount && user.failedLoginCount >= options.failedLoginCount) {
       seneca.log.debug('login/fail', why = 'locked-out', user)
       return done(null, {ok: false, why: why, user: user})
```
Comment (mirceaalexandru) [inline user.js]: Here the user is returned as part of the response. Because the `ok: false` I think user should not be returned.

File: user.js [hunk 5]
```diff
@@ -645,7 +645,7 @@ module.exports = function user (options) {
         if (err) return done(err)
         if (!out.ok) {
           seneca.log.debug('login/fail', why = 'invalid-password', user)
-          seneca.act({role: role, cmd: 'set_user_lock', id: user.id, failTry: true, why: why}, function (err, out) {
+          cmd_increment_lock(seneca, user.id, false, function (err, out) {
             if (err) return done(err)
             done(null, {ok: false, why: why, user: user})
```
Comment (mirceaalexandru) [inline user.js]: Same here, I think user should not be returned in case ok is false.

File: user.js [hunk 6]
```diff
@@ -675,49 +675,55 @@ module.exports = function user (options) {
 
       login.save$(function (err, login) {
         if (err) return done(err)
-        seneca.act({role: role, cmd: 'set_user_lock', id: user.id}, function (err, out) {
+        cmd_increment_lock(seneca, user.id, true, function (err, out) {
           if (err) return done(err)
+
           seneca.log.debug('login/ok', why, user, login)
           done(null, {ok: true, user: user, login: login, why: why})
```
Comment (mirceaalexandru) [inline user.js]: I think we should return user only after cleaning the information - take a look on `cleanUser` function.

I think that on `cleanUser` we should also remove the retry failed count information as we should not make it public/return it in the response.

File: user.js [hunk 7]
```diff
@@ -675,49 +675,55 @@ module.exports = function user (options) {
...
-        ok = true
-        why = 'account-unlocked'
+        user.failedLoginCount++
       }
       user.save$(function (err, user) {
         if (err) return done(err, {ok: false, why: err})
-        seneca.log.debug('user/lock', why, user)
-        done(null, {ok: ok, why: why})
+        seneca.log.debug('increment lock ', user, 'failedLoginCount ', user.failedLoginCount)
+        done(null, {ok: true, why: 'login-count-modified'})
```
Comment (mirceaalexandru) [inline user.js]: Maybe here we don't want to return a `why` as the `ok` is true.

File: user.js [hunk 8]
```diff
@@ -675,49 +675,55 @@ module.exports = function user (options) {
...
+  // Unlocks accounts that a user has been locked out of due to failed password attempts
+  // - id: user id to unlock
+  // Provides:
+  // - success: {ok:true}
+  // - failure: {ok:false}
+  function cmd_unlock (args, done) {
+    var seneca = this
+    cmd_increment_lock(seneca, args.id, true, function (err, out) {
+      if (err) return done(err, {ok: false, why: err})
+      done(null, {ok: true, why: 'account-unlocked'})
```
Comment (mirceaalexandru) [inline user.js]: same as above?

File: user.js (hunk 6 above)
Comment (mihaidma) [inline user.js]: added fail count in the answer just for info purposes, thinking that it is not security critical info. I could remove it although I don't see the issue with it

Comment (mihaidma) [inline user.js]: i'll make a separate task related to cleaning returned data as it needs more detailed full code check. That code area is not related to changes made by the brute force implementation

File: test/lock-unlock.test.js (hunk 1 above)
Comment (mihaidma) [inline test/lock-unlock.test.js]: id is set later after the account is created in init

**Pull request merged**